monero-sys/src/lib.rs

```diff
@@ -0,0 +1,1511 @@
+    fn start_refresh(&mut self) {
+        self.inner.pinned().startRefresh();
+    }
+
+    /// Refresh the wallet asynchronously.
+    /// Same as start_refresh except that the background thread only
+    /// refreshes once. Maybe?
+    fn refresh_async(&mut self) {
+        self.inner.pinned().refreshAsync();
+    }
+
+    /// Refresh the wallet synchronously.
+    /// No possibility for progress reporting.
+    fn refresh_blocking(&mut self) -> anyhow::Result<()> {
+        let success = self.inner.pinned().refresh();
+
+        if !success {
+            let connected = self.connected();
+            tracing::error!(connected, "Failed to sync Monero wallet");
+            self.check_error().context("Failed to refresh wallet")?;
+            anyhow::bail!("Failed to refresh wallet (no reason given)");
+        }
+
+        Ok(())
+    }
+
+    /// Get the wallet creation height.
+    fn creation_height(&self) -> u64 {
+        self.inner.getRefreshFromBlockHeight()
+    }
+
+    /// Get the current blockchain height.
+    fn blockchain_height(&self) -> u64 {
+        self.inner.blockChainHeight()
+    }
+
+    /// Get the daemon's blockchain height.
+    ///
+    /// Returns the height of the blockchain, if connected.
+    /// Returns None if not connected.
+    fn daemon_blockchain_height(&self) -> Option<u64> {
+        tracing::debug!(connected=%self.connected(), "Getting daemon blockchain height");
+
+        // Here we actually use the _target_ height -- incase the remote node is
+        // currently catching up we want to work with the height it ends up at.
+        match self.inner.daemonBlockChainTargetHeight() {
+            0 => None,
+            height => Some(height),
+        }
+    }
+
+    /// Get the total balance across all accounts.
+    fn total_balance(&mut self) -> monero::Amount {
+        let balance = self.inner.balanceAll();
+        monero::Amount::from_pico(balance)
+    }
+
+    /// Get the total unlocked balance across all accounts in atomic units.
+    fn unlocked_balance(&mut self) -> monero::Amount {
+        let balance = self.inner.unlockedBalanceAll();
+        monero::Amount::from_pico(balance)
+    }
+
+    /// Check if the wallet is synced with the daemon.
+    ///
+    /// The original implementation trusted `wallet->synchronized()` from the C++
+    /// layer, but in practice this flag is set after the very first refresh and
+    /// does **not** take newly mined blocks into account. This led to false
+    /// positives where the Rust side considered the wallet synced even though
+    /// it still had to scan additional blocks, resulting in `balance == 0`.
+    ///
+    /// We now regard the wallet as synced **only** when the wallet's perceived
+    /// height has caught up with the daemon height.  In regtest environments
+    /// the wallet may occasionally overshoot the target height, therefore we
+    /// accept `current >= target`.
+    fn synchronized(&self) -> bool {
```

Comment:
I think `wallet->synchronized()` may be much better suited for production usage as it takes into account wallet state and edge cases we might not know about.

I think the only reason this doesn't work as well in regtest is because of the fast block times?